File: sdk/csharp/src/AnyVali/Schemas/NullSchema.cs
namespace AnyVali.Schemas;

public sealed class NullSchema : Schema
{
    internal override object? Validate(object? input, ValidationContext ctx)
    {
        if (input is not null)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.InvalidType,
                Message = $"Expected null, received {DescribeType(input)}",
                Path = ctx.ClonePath(),
                Expected = "null",
                Received = DescribeType(input),
            });
            return null;
        }
        return null;
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?> { ["kind"] = "null" };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new NullSchema
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}

File: sdk/ruby/lib/anyvali/schemas/null_schema.rb
# frozen_string_literal: true

module AnyVali
  class NullSchema < Schema
    def initialize(**kwargs)
      super(kind: "null", **kwargs)
    end

    protected

    def validate(value, path, issues, context)
      unless value.nil?
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: "null",
          received: Schema.type_name(value)
        )
      end
    end
  end
end
